plugins/inputs/turbostat/parser/parser.go

```diff
@@ -0,0 +1,232 @@
+	'%': {},
+	'+': {},
+	'-': {},
+}
+
+var knownTokens = []string{
+	"Watt",
+	"MHz",
+	"Tmp",
+	"Thr",
+	"GFX",
+	"SAM",
+}
+
+var tagNames = map[string]struct{}{
+	"package": {},
+	"die":     {},
+	"core":    {},
+	"cpu":     {},
+	"apic":    {},
+	"x2apic":  {},
+}
+
+type column struct {
+	name       string
+	isTag      bool
+	isIgnored  bool
+	multiplier float64
+}
+
+type tagMap = map[string]string
+type fieldMap = map[string]any
+
+// Reads metrics from a stream (such as Turbostat stdout) and adds them
+// to an accumulator. If an error is encountered, the function returns it
+// and stops further processing.
+func ProcessStream(r io.Reader, acc telegraf.Accumulator) error {
+	scanner := bufio.NewScanner(r)
+	if !scanner.Scan() {
+		return scanner.Err()
+	}
+	header := scanner.Text()
+	columns := make([]column, 0)
+	for _, s := range strings.Fields(header) {
+		columns = append(columns, createColumn(s))
+	}
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == header {
+			continue
+		}
+		values := strings.Fields(line)
+		tags, fields, err := processValues(columns, values)
+		if err != nil {
+			return err
+		}
+		acc.AddFields("turbostat", fields, tags)
+	}
+	return scanner.Err()
+}
+
+// Creates tags and fields from columns and values.
+func processValues(columns []column, values []string) (tagMap, fieldMap, error) {
+	if len(values) > len(columns) {
+		msg := "too many values: %d columns, %d values"
+		err := fmt.Errorf(msg, len(columns), len(values))
+		return nil, nil, err
+	}
+	tags := make(tagMap, 0)
+	fields := make(fieldMap, 0)
+	for i := range values {
+		if columns[i].isIgnored {
+			continue
+		}
+		if columns[i].isTag {
+			if !isTagValue(values[i]) {
+				return nil, nil, fmt.Errorf("invalid tag: %s", values[i])
+			}
+			tags[columns[i].name] = values[i]
+		} else {
+			v, err := strconv.ParseFloat(values[i], 64)
+			if err != nil {
+				return nil, nil, err
+			}
+			if columns[i].multiplier != 0 {
+				v *= columns[i].multiplier
+			}
+			fields[columns[i].name] = v
+		}
+	}
+	if len(fields) == 0 {
+		return nil, nil, errors.New("no value for any field")
+	}
+	return tags, fields, nil
+}
+
+// Creates a column struct from a Turbostat column name.
+func createColumn(s string) column {
+	c := column{}
+	// Split the Turbostat column name into tokens.
+	tokens := make([]string, 0)
+	for _, element := range splitSymbols(s) {
+		tokens = append(tokens, splitKnownTokens(element)...)
+	}
+	for i, token := range tokens {
+		token = strings.ToLower(token)
+		// If a tokens is "mhz", set the multiplier to 1M.
+		if token == "mhz" {
+			c.multiplier = 1e6
+		}
+		// If a token is an abbrebiation, replace it.
+		if replacement, ok := abbreviations[token]; ok {
+			token = replacement
+		}
+		tokens[i] = token
+	}
+	// Build the camel case column name.
+	c.name = strings.Join(tokens, "_")
+	// Ignore the timestamp column.
+	if c.name == "time_of_day_seconds" {
+		c.name = "timestamp"
+		c.isIgnored = true
+	}
```

Comment:
Since this column is ignored anyway, changing its name might be a bit misleading...

Reply:
Removed the rename, thanks.